modules/src/ics02_client/handler/create_client.rs

```diff
@@ -0,0 +1,286 @@
+        fn client_type(&self) -> ClientType {
+            todo!()
+        }
+
+        fn height(&self) -> Height {
+            todo!()
+        }
+    }
+
+    #[derive(Copy, Clone, Debug, PartialEq, Eq)]
+    struct MockClientState(u32);
+
+    impl ClientState for MockClientState {
+        type ValidationError = MockError;
+
+        fn client_id(&self) -> ClientId {
+            todo!()
+        }
+
+        fn client_type(&self) -> ClientType {
+            todo!()
+        }
+
+        fn get_latest_height(&self) -> Height {
+            todo!()
+        }
+
+        fn is_frozen(&self) -> bool {
+            todo!()
+        }
+
+        fn verify_client_consensus_state(
+            &self,
+            _root: &CommitmentRoot,
+        ) -> Result<(), Self::ValidationError> {
+            todo!()
+        }
+    }
+
+    impl From<MockConsensusState> for MockClientState {
+        fn from(cs: MockConsensusState) -> Self {
+            Self(cs.0)
+        }
+    }
+
+    #[derive(Copy, Clone, Debug, PartialEq, Eq)]
+    struct MockConsensusState(u32);
+
+    impl ConsensusState for MockConsensusState {
+        type ValidationError = MockError;
+
+        fn client_type(&self) -> ClientType {
+            todo!()
+        }
+
+        fn height(&self) -> Height {
+            todo!()
+        }
+
+        fn root(&self) -> &CommitmentRoot {
+            todo!()
+        }
+
+        fn validate_basic(&self) -> Result<(), Self::ValidationError> {
+            todo!()
+        }
+    }
+
+    #[derive(Clone, Copy, Debug, PartialEq, Eq)]
+    struct MockClient;
+
+    impl ClientDef for MockClient {
+        type Header = MockHeader;
+        type ClientState = MockClientState;
+        type ConsensusState = MockConsensusState;
+    }
+
+    #[derive(Clone, Debug, PartialEq, Eq)]
+    struct MockClientContext {
+        client_id: ClientId,
+        client_state: Option<MockClientState>,
+        client_type: Option<ClientType>,
+        consensus_state: Option<MockConsensusState>,
+    }
+
+    impl ClientContext<MockClient> for MockClientContext {
+        fn client_type(&self, client_id: &ClientId) -> Option<ClientType> {
+            if client_id == &self.client_id {
+                self.client_type.clone()
+            } else {
+                None
+            }
+        }
+
+        fn client_state(&self, client_id: &ClientId) -> Option<MockClientState> {
+            if client_id == &self.client_id {
+                self.client_state
+            } else {
+                None
+            }
+        }
+
+        fn consensus_state(
+            &self,
+            client_id: &ClientId,
+            _height: Height,
+        ) -> Option<MockConsensusState> {
+            if client_id == &self.client_id {
+                self.consensus_state
+            } else {
+                None
+            }
+        }
+    }
+
+    #[test]
+    fn test_create_client_ok() {
+        let mock = MockClientContext {
```

Comment:
As shown below, this is missing the `client_id`. 
What is the expectation on the caller that needs to give a context to `process()` in addition to the message? Specifically for clients, if the client does not exist, the context should still include a valid `client_id` and this test passes as long as one is given. Outside tests, is the caller expected to check if this is a `MsgCreateClient` create a context with `None` for state and type if the client doesn't exist or return existing if it does? We briefly talked about this last week and this is still an aspect we should clarify. For connections the context/ keeper passed to `process()` needs both client and connection context, for channels we need client, connection and channel, etc. Things will become more complex when setting up all the mocks for tests.

But the bigger concern is that the IBC handler (as per ICS) logic is split somewhat between `handle()` and the caller.

Reply:
The expectations are captured by the `CientContext` trait:

```rust
pub trait ClientContext<CD>
where
    CD: ClientDef,
{
    fn client_type(&self, client_id: &ClientId) -> Option<ClientType>;
    fn client_state(&self, client_id: &ClientId) -> Option<CD::ClientState>;
    fn consensus_state(&self, client_id: &ClientId, height: Height) -> Option<CD::ConsensusState>;
}
```

The fact that the mock requires a client id is an implementation detail of this particular mock struct.
One could imagine implementing the `ClientContext` trait in a different way, but the way it's done here is imho pretty straightforward as one can exercise all code paths in the `process` function just by constructing different values of the mock struct.

> But the bigger concern is that the IBC handler (as per ICS) logic is split somewhat between handle() and the caller.

Could you expand on that, in the context of what I wrote above, if that still makes sense?